torch/utils/data/graph.py

```diff
@@ -46,14 +52,23 @@ def reduce_hook(obj):
             cls.set_reduce_ex_hook(reduce_hook)
             if only_datapipe:
                 cls.set_getstate_hook(getstate_hook)
-        p.dump(scan_obj)
+        try:
+            p.dump(scan_obj)
+        except (pickle.PickleError, AttributeError, TypeError):
+            if DILL_AVAILABLE:
+                d.dump(scan_obj)
+            else:
+                pass
```

Comment:
Wait, I don't think we should pass here. If the `DataPipe` can not be `dump`ed, we can't save it to the graph. So, a proper Error should raise to indicate that.

Reply:
You are right, I am changing `pass` to `raise`

Reply:
I suspect it may be caught by the next line (62-63), unless we want to change that right now

Reply:
I think we should remove 62-63 as well. wdyt

Reply:
Yea, I think it is better for the users/us to know if something is not pickable rather than silently failing and not capturing it within the graph.

Reply:
IIRC, this traverse function is only used by `DataLoader2`? If `DataPipe` is provided to `DataLoader`, it won't check if the `DataPipe` graph is traversable or not. A proper error would raise when multiprocessing involved.

Reply:
`traverse` is used by `apply_shuffle_settings`, which is used here within DataLoader:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L245

Reply:
Thank you for clarification. I still think it's better to remove the logic to suppressing pickling Error as you pointed out.

> I think it is better for the users/us to know if something is not pickable rather than silently failing and not capturing it within the graph.

Reply:
Agree, I am updating this now.